src/Collection/CollectionTrait.php

```diff
@@ -686,6 +686,58 @@ public function _unwrap()
      *
      * @return \Cake\Collection\CollectionInterface
      */
+    public function cartesianProduct(callable $operation = null, callable $filter = null)
+    {
+        if ($this->isEmpty()) {
+            return new Collection([]);
+        }
+
+        $collectionArrays = [];
+        $collectionArraysKeys = [];
+        $collectionArraysCounts = [];
+
+        foreach ($this as $value) {
+            if (count($value) !== count($value, COUNT_RECURSIVE)) {
+                throw new LogicException('Cannot find the cartesian product of a multidimensional array');
+            }
+
+            $collectionArraysKeys[] = array_keys($value);
+            $collectionArraysCounts[] = count($value);
+            $collectionArrays[] = $value;
+        }
+
+        $result = [];
+        $lastIndex = count($collectionArrays) - 1;
+        // holds the indexes of the arrays that generate the current combination
+        $currentIndexes = array_fill(0, $lastIndex + 1, 0);
+
+        $changeIndex = $lastIndex;
+
+        while (!($changeIndex === 0 && $currentIndexes[0] === $collectionArraysCounts[0])) {
+            $currentCombination = array_map(function ($value, $keys, $index) {
+                return $value[$keys[$index]];
+            }, $collectionArrays, $collectionArraysKeys, $currentIndexes);
+
+            if ($filter === null || call_user_func_array($filter, $currentCombination)) {
```

Comment:
I think making each value in the combination a function parameter makes this really difficult to use. Why not just pass an array as a single argument to the function?